File: brady/best_time_to_buy_and_sell_stock.java
class Solution {
  
    // O(n) time, O(1) space
    public int maxProfit(int[] prices) {
        int minVal = prices[0];
        int minIndex = 0;
        int maxProfit = 0;
        for (int i = 1; i < prices.length; ++i) {
            if (prices[i] < minVal) {
                minVal = prices[i];
                minIndex = i;
            } else {
                maxProfit = Math.max(maxProfit, prices[i] - minVal);
            }
        }
        return maxProfit;
    }
}

File: manraj/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy = 0
        max_profit = 0
        sell = 1

        while sell < len(prices):
            if prices[buy] < prices[sell]:
                max_profit = max(prices[sell] - prices[buy], max_profit)
            else:
                buy = sell

            sell += 1

        return max_profit

